### PR TITLE
Server: Implement a RootHandler to `/`

### PR DESCRIPTION
**Pull request by danielfrg**
Implement a RootHandler to `/` that lists available applications or redirects

- [x] issues: fixes https://github.com/bokeh/bokeh/issues/4412
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by danielfrg** [comment]
Can someone take a look and see if the failing test build is because of the PR?. I don't think so. If i am correct can someone rerun the build?

**Comment by bryevdv** [comment]
@danielfrg I have restarted. Been seeing alot of spurious issues with selenium integration lately, willl have to investigate separately. Agreed it has nothing to do with your PR (thanks BTW!)

**Comment by bryevdv** [comment]
Does it make sense to add a command line option to enable/disable the addition of the `RootHandler` ?

**Comment by danielfrg** [comment]
I think this a nicer behaviour than the current default. Going to `localhost:8080` and seeing a 404 is kinda disappointing and there is not a big log that tells you were the app is running. I can add a flag to disable if you think is necessary.

**Comment by bryevdv** [comment]
Oh, yes sorry I think this should be the default, but I can also imagine in some kind of deployments that users would like the ability to explicitly disable it as well. Perhaps `--disable-index` to turn it off?

**Comment by bryevdv** [comment]
FYI @danielfrg I think the failing test is because the Sauce Labs credentials are only made available to the tests job if you are in @bokeh/dev I thought there would be a warning in this case but apparently it is failing. We will have to address this, but your PR can be merged as soon as all other tests suites pass, and the flag is added. Thanks again!

**Comment by danielfrg** [comment]
@bryevdv I added the flag. The same build is still failing. Let me know if there is something else I should do.

**Comment by bryevdv** [comment]
Hi @danielfrg I tested this out and it works great. I am happy to merge this as-is, but if you have the time and interest, there are two other things that could be done:

##### make the index page more attractive 

The index might have a nice header with logo, a minimal paragraph describing the Bokeh server with links to the docs, and an explanation of what's listed below, etc. The list appearance itself could be improved with some CSS styling, or even presented with something like a Bootstrap list group.

##### add some minimal documentation

The command line option can be described in the `server.py` docstring, and then it will show up in the user guide section on the "command line tools". Once the index page is prettied up some, it would also be good to add a small narrative section with screenshot to the users guide section on the server.

----

I completely understand if you are busy, etc. so it's really up to you. We can add issues for someone to do this as follow-on work in other PRs just as well. Either way, thanks very much for the nice PR!

File: bokeh/server/views/root_handler.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    @gen.coroutine
+    def get(self, *args, **kwargs):
+        if len(self.applications) == 1:
+            app_names = list(self.applications.keys())
+            redirect_to = (self.prefix if self.prefix else "") + app_names[0]
+            self.redirect(redirect_to)
+        else:
+            self.write("<ul>")
+            for app_route, _ in self.applications.items():
+                self.write("<li><a href={0}>{0}</a></li>".format(app_route))
```
Comment (havocp) [inline bokeh/server/views/root_handler.py]: the app's document.title could be used here for the link text, which might be prettier than the raw app route.

**Comment by danielfrg** [comment]
@bryevdv Thanks for reviewing and testing. I am interested on documenting and having a nicer landing page but I will be out for a week and I would like to have this merged before that so I can use it on a something I am building. I could use my branch but its a little annoying to deploy.

If its with you i say lets merge and create the issues and ping me on those and I'll take care of than when I am back.

**Comment by bryevdv** [comment]
@havocp that's not a bad idea, I've made not of it in #4425 which will be a follow-on to improve and refine this functionality. 

@danielfrg I will go ahead and merge this, FYI we probably won't advertise this feature too widely until #4425 is done, which may or may not be before `0.12`

**Pull request merged**